library/goptions.mli

```diff
@@ -131,18 +138,45 @@ type 'a option_sig = {
 (** The [preprocess] function is triggered before setting the option. It can be
     used to emit a warning on certain values, and clean-up the final value.
 
-    [declare_stringopt_option] should be preferred to [declare_string_option]
-    because it supports "Unset".
-    Only "Warnings" option is declared using the latter.*)
+    [StringOptKind] should be preferred to [StringKind] because it supports "Unset". *)
+val declare_option   : ?preprocess:('a -> 'a) -> 'a option_sig -> unit
+
+val declare_append_only_option : ?preprocess:(string -> string) -> sep:string ->
+  string option_sig -> unit
+
+(** Unsynchronized options *)
+module RawOpt : sig
+  type 'a t = {
+    kind : 'a option_kind;
+    depr : Deprecation.t option;
+    stage : Summary.Stage.t;
+    read : unit -> 'a;
+    write : option_locality -> 'a -> unit;
+  }
+end
 
-val declare_int_option   : ?preprocess:(int option -> int option) ->
-                           int option option_sig -> unit
-val declare_bool_option  : ?preprocess:(bool -> bool) ->
-                           bool option_sig   -> unit
-val declare_string_option: ?preprocess:(string -> string) ->
-                           string option_sig -> unit
-val declare_stringopt_option: ?preprocess:(string option -> string option) ->
-                              string option option_sig -> unit
+(** Declare an unsynchronized option.
+    This means that [write] is called directly by "Set Foo" (after type casting).
+
+    This is in particular useful to define options which wrap another option,
+    eg suppose "Foo Debug Verbosity" is from [declare_intopt_option_and_ref]
+    and we want to declare a flag "Foo Debug" which when unset will unset "Foo Debug Verbosity",
+    and when set will set "Foo Debug Verbosity" to 1.
+    If we used [declare_option], backtracking would reset "Foo Debug Verbosity" and "Foo Debug"
+    in an undefined order, such that if "Foo Debug Verbosity" was 2
+    if "Foo Debug" is reset later it gets overwritted to 1.
+    ie "Set Foo Debug Verbosity 2. Something unrelated. Undo 1." may result in
+    "Foo Debug Verbosity = 1".
+
+    Therefore instead we should declare "Foo Debug" as an unsynchronized option, which calls
+    [set_int_option_value_gen] on "Foo Debug Verbosity"
+    (NB: the stage in [RawOpt.t] controls in which phase [write] is called).
+*)
+val declare_unsynchronized_option : option_name -> 'a RawOpt.t -> unit
+
+(** Helper to build synchronized options on top of [declare_unsynchronized_option]. *)
+val option_object : string -> Summary.Stage.t -> ('a -> unit) ->
+  (option_locality * 'b, option_locality * 'a, option_locality * 'b) Libobject.object_declaration
```

Comment:
@SkySkimmer , can we find a different solution for this problem?

This will break in coq-lsp / Flèche for example, and it makes impossible to properly cache command execution. Moreover, I think this is generally a footgun.

Note that we used to have this kind of "unsynchronized" options but there were removed long time ago, happy to dig the relevant issues PRs if you are curious. IIRC, all uses of them were incorrect.

Reply:
I think for the problem in https://github.com/MetaCoq/metacoq/pull/1123 an unsynchronized option (but backed by the synchronized state of the other option) really is the correct solution.

I don't know what you mean by "this will break"

Reply:
I mean:

- assume an unsynchronized option "Foo." that affects for example de-elaboration.
- imagine that I have a file:

```coq
Theorem foo : ...

Set Foo.

Theorem bar : ...
```

- now the user clicks on bar, so I print the goals there
- now the user clicks on foo, so I restore the state to print the goals there
- however the option is still set as lives outside the summary.

I do think we have a need for this kind of use cases, for example, I don't like that backtracking setting lives in the summary, however IMHO we should not expose via GOptions, see the long discussion when we removed the unsync parameter.

What do you think? We can maybe split that change to a different PR while we discuss if you wanna get the rest merged.

Reply:
The way I think of it, declare_unsynch_option is mostly not for declaring actually unsynchronized options, rather it's for declaring options with different synchronization from what the standard declare_option does.

Reply:
What is "different synchronization option from what the standard declare_option does" mean? Isn't that "unsynchronized"?

Reply:
cf the comment on val declare_unsynchronized_option

Reply:
I must be missing something, the comment says they are the same.

Reply:
The option_object helper is the same (I added it when I was thinking about implementing Warnings/Debug in a more adhoc way outside this file, in the end I did declare_append_only_option in this file so it could be private)
but I thought we were talking about declare_unsynchronized_option

Reply:
I'm talking about declare_unsynchronized_option indeed, sorry I placed my comment on the wrong place.

Let me re-state my worry: I'm against `declare_unsynchronized_option` for the reasons stated above.

Would it be too much of a trouble to split that change to a different PR and continue the discussion there?

I see `declare_unsynchronized_option` is not used here, and the rest of the PR seems fine.

Reply:
It does seem like quite a bad idea to break the well-established existing behavior when rolling back options.  I would ask a) is this only possible way to fix this bug? and b) what are other examples of when this feature would be useful?  It is really a design flaw in metacoq?  Unsynchronized options seems like something that's ripe for abuse/problems and perhaps hard to find when there is a problem.

Reply:
>Would it be too much of a trouble to split that change to a different PR and continue the discussion there?

unsynchronized_option and being able to fix the metacoq bug instead of having to delete the feature is the main motivation for the PR, the other changes are mainly to make it nicer (eg not having to deal with append)

I don't know what is not clear about the need for unsynchronized_option in the comment above it.

Reply:
It is not clear why we need to introduce stuff that does break live the summary and as Jim points out will break expected document behavior.

Also it is not clear why we cannot use a different solution, like adding grouping or ordering instead of this in my opinion very problematic solution with large potential for misuse.

Reply:
It doesn't break the summary, declare_option isn't even where the state is declared (unlike declare_option_and_ref)